c10/core/DispatchKey.h

```diff
@@ -19,47 +27,85 @@ enum class DispatchKey : uint8_t {
-  // dispatching on at::Generator* parameters.
-  // This key must be used in two places:
-  //  1) as a second parameter of at::Generator constructor call in
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~ BACKENDS ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
+  // A "backend" is colloquially used to refer to handlers for dispatch
+  // which actually implement the numerics of an operation in question.
+  // Alternately, it is placed in the "back" of the dispatch hierarchy,
+  // as it is handled last and is a terminal handler.
+  //
+  // The set of backends is not really ordered (except in the case of
+  // sparse), but as our model of dispatching requires every dispatch key
+  // to participate in a total order, we have an unused ordering here.)
+
+  // Here are backends which you think of as traditionally specifying
+  // how to implement operations on some device.
+  CPUTensorId,    // registered at build/aten/src/ATen/CPUType.cpp
+  CUDATensorId,   // registered at build/aten/src/ATen/CUDAType.cpp
+  HIPTensorId,    // TODO: I think this is not actually used, due to Note [Masquerading as CUDA]
+  MSNPUTensorId,  // tested at test/cpp_extensions/msnpu_extension.cpp
+                  // NB: we suspect that MSNPU is defunct
+  XLATensorId,    // lives out of tree at https://github.com/pytorch/xla
+
+  // These are Caffe2 device types which we grandfathered into
+  // DispatchKey.
+  // TODO: Caffe2-only DispatchKeys actually should be removed from this enum
+  // and just simply be undispatchable.
+  MKLDNNTensorId, // (MKLDNN is treated as another "device" in Caffe2)
+  OpenGLTensorId,
+  OpenCLTensorId,
+  IDEEPTensorId,
+
+  // Here are backends which specify more specialized operators
+  // based on the dtype of the tensor.
+  QuantizedCPUTensorId, // registered at build/aten/src/ATen/QuantizedCPUType.cpp
+  ComplexCPUTensorId,   // lives out of tree at https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex
+  ComplexCUDATensorId,  // and https://gitlab.com/pytorch-complex/pytorch-cuda-strided-complex
+                        // tested at test/cpp_extensions/complex_registration_extension.cpp
+                        // TODO: Remove Complex dispatch keys when Complex is moved in tree
+
+  // This backend is to support custom RNGs; it lets you go
+  // to a different kernel if you pass in a generator that is not a
+  // traditional CPUGenerator/CUDAGenerator.  To make use of this
+  // key:
+  //  1) set it as a second parameter of at::Generator constructor call in
   //     the user-defined PRNG class.
-  //  2) as a dispatch key while registering custom kernels
+  //  2) use it as a dispatch key while registering custom kernels
   //     (templatized kernels specialized for user-defined PRNG class)
+  // intended for out of tree use; tested by aten/src/ATen/test/rng_test.cpp
   CustomRNGKeyId,
 
+  // Here are backends which specify more specialized operators
+  // based on the layout of the tensor.  Note that the sparse backends
```

Comment:
ultra nit: a few occasions of double spacing.

Reply:
It's the old type writer style where you double space after a period :>